### PR TITLE
Add database and handle users

### PR DESCRIPTION
**Pull request by dandds**
Did this one with @richard-dds 

This does a few important things for authnid:

- adds a Postgres database with a basic user schema
- uses genuine UUIDs in the bearer tokens
- the `validate` endpoint now returns user data for ATST based on the UUID in the bearer token it's given

When a user hits the root login endpoint:
- we parse their CAC info from the cert
- we check the database to see if it matches an existing user
    - if yes, we use their existing UUID to generate the bearer token
    - if no, we add them to the database and use the new UUID to generate the bearer token
- we redirect them to ATST with the bearer-token, as before

When ATST queries the `validate` endpoint with a bearer token:
- we validate the bearer token as before
- if the token is valid, we return a JSON response that includes the user data

### CODE REVIEW

File: authnid/make_db.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    return connection.cursor()
+
+def make_db(config):
+    connection = connect_db(config['DATABASE_URI'])
+    cursor = make_cursor(connection)
+    cursor.execute('''
+        CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+        CREATE TABLE IF NOT EXISTS users (
+            id uuid DEFAULT uuid_generate_v4() PRIMARY KEY,
+            email VARCHAR(254),
```
Comment (briandds) [inline authnid/make_db.py]: Is there an advantage to specifying the length of these, rather than just `varchar`?

Comment (dandds) [reply authnid/make_db.py]: Probably not. Just trying to be specific. Do you think that's asking for trouble?

Comment (jordangov) [reply authnid/make_db.py]: Yeah, we don't need that level of optimization.

File: authnid/make_app.py [hunk 2]
```diff
@@ -65,3 +78,18 @@ def _make_crl_validator(app):
 
 def _make_token_manager(app):
     app.token_manager = TokenManager(app.config.get("TOKEN_SECRET"))
+
+
+def _apply_request_hooks(app):
+    @app.before_request
+    def before_request():
+        app.db = connect_db(app.db_uri)
```
Comment (jordangov) [inline authnid/make_app.py:84]: Does this do connection pooling? Or am I prematurely optimizing with my thoughts? My point being... are we spinning up a new DB connection on each request when we don't need to?

File: authnid/make_db.py (hunk 1 above)
Comment (dandds) [reply authnid/make_db.py]: Okay. Removed those and added autocommit on two new commits.

File: authnid/make_app.py [hunk 3]
```diff
@@ -65,3 +78,18 @@ def _make_crl_validator(app):
...
+        cursor = make_cursor(app.db)
+        autocommit = app.config['ENV'] != 'test'
```
Comment (jordangov) [inline authnid/make_app.py]: This is just my inexperience talking, but why don't we want test data to commit? Do we not have any tests that would require data being committed and then checked at another point? I guess I would expect the DB table(s) to be rebuilt for test runs (versus not committing data).

File: authnid/make_app.py (hunk 2 above)
Comment (dandds) [reply authnid/make_app.py:84]: No connection pooling right now. This was just sort of a rough-out. Right now it is per-request, but we could do it on app spin-up and teardown.

File: authnid/make_db.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+def make_cursor(connection):
+    return connection.cursor()
+
+def make_db(config):
+    connection = connect_db(config['DATABASE_URI'])
+    cursor = make_cursor(connection)
+    cursor.execute('''
+        CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+        CREATE TABLE IF NOT EXISTS users (
```
Comment (jordangov) [inline authnid/make_db.py]: I get that we're still in rough mode, but I strongly feel that implementing DB migrations now will make things much smoother going forward. I strongly recommend going that route now versus starting with hard-coded schemas.

File: authnid/make_db.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            id uuid DEFAULT uuid_generate_v4() PRIMARY KEY,
+            email VARCHAR,
```
Comment (jordangov) [inline authnid/make_db.py]: unique?

File: authnid/user_repo.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        return self.cursor.mogrify(
+                    self.IS_MATCH,
+                    (AsIs(','.join(kwargs.keys())), tuple(kwargs.values()))
+                ).decode()
+
+    def has_user(self, **kwargs):
+        match = self._build_match(**kwargs)
+        self.cursor.execute("""
+        SELECT CASE WHEN EXISTS (
+            SELECT * FROM users WHERE {}
```
Comment (jordangov) [inline authnid/user_repo.py]: `SELECT *` is not safe, and will not pass (should not pass) security review. Remember: Zero Trust. Only select the fields you need. We can always add more.

File: authnid/user_repo.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        )
+        THEN true
+        ELSE false END
+        """.format(match))
+        return self.cursor.fetchone()[0]
+
+    def add_user(self, email=None, dod_id=None, first_name=None, last_name=None):
```
Comment (jordangov) [inline authnid/user_repo.py]: getting kinda long on the args, why not a hash?

File: authnid/user_repo.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        INSERT INTO users (email,dod_id,first_name,last_name)
+            VALUES (%s,%s,%s,%s)
+            RETURNING id;
+        """, (email, dod_id, first_name, last_name))
+        uuid = self.cursor.fetchone()[0]
+        return uuid
+
+    def get_user(self, **kwargs):
+        match = self._build_match(**kwargs)
+        self.cursor.execute("SELECT * FROM users WHERE {}".format(match))
```
Comment (jordangov) [inline authnid/user_repo.py]: Another `SELECT *` ... please change.

File: authnid/user_repo.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        return self.cursor.fetchone()
+
+    def count(self):
+        self.cursor.execute("SELECT count(*) FROM users")
```
Comment (jordangov) [inline authnid/user_repo.py]: We should just be able to count the UUID field, right?

File: authnid/user_repo.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        return self.cursor.fetchone()[0]
+
+    def all(self):
+        self.cursor.execute("SELECT * FROM users")
```
Comment (jordangov) [inline authnid/user_repo.py]: Another `SELECT *` ... please change.

File: authnid/user_repo.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    def get_user(self, **kwargs):
+        match = self._build_match(**kwargs)
+        self.cursor.execute("SELECT * FROM users WHERE {}".format(match))
+        return self.cursor.fetchone()
+
+    def count(self):
+        self.cursor.execute("SELECT count(*) FROM users")
+        return self.cursor.fetchone()[0]
+
+    def all(self):
```
Comment (jordangov) [inline authnid/user_repo.py]: Hmm... are we using this method? If not, we need to remove this, it's a giant security hole.

**Review by jordangov — CHANGES_REQUESTED** [review]

File: authnid/make_db.py (hunk 5 above)
Comment (dandds) [reply authnid/make_db.py]: Would DOD ID also be unique? And are we certain there are no legit cases where we'd have redundant emails? Like if someone were issued a new DOD ID but their email doesn't change?

**Comment by dandds** [comment]
@jordangov The recent commits should cover most of your feedback. 

Do you think connection pooling should be on this PR, or can that wait for a later one?

**Review by richard-dds — APPROVED** [review]

**Comment by jordangov** [comment]
I'm fine without pooling for this go around, but let's add it to the icebox. I'll need to review these changes late tonight or tomorrow as I'm traveling today.

**Comment by dandds** [comment]
Connection pooling PT chore: https://www.pivotaltracker.com/story/show/158723960

File: alembic.ini [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+# version location specification; this defaults
+# to alembic/versions.  When using multiple version
+# directories, initial revisions must be specified with --version-path
+# version_locations = %(here)s/bar %(here)s/bat alembic/versions
+
+# the output encoding used when revision files
+# are written from script.py.mako
+# output_encoding = utf-8
+
+sqlalchemy.url = driver://user:pass@localhost/dbname
```
Comment (jordangov) [inline alembic.ini:38]: Not familiar with Alembic, but is this just a template? Seems like it wouldn't be our URL. :)

File: alembic/env.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+# Add root project dir to the python path
+import sys
+from unipath import Path
+
+parent_dir = Path(__file__).parent.parent
+sys.path.append(parent_dir)
+
+from authnid.make_app import make_config
+app_config = make_config()
+config.set_main_option('sqlalchemy.url', app_config['DATABASE_URI'])
```
Comment (jordangov) [inline alembic/env.py:27]: Ah... this may address my other line comment about the DB URL?

Comment (dandds) [reply alembic/env.py:27]: @richard-dds is more familiar with Alembic config than I am, but that's my understanding, yeah. We just override it here from our default config so we don't have to worry about another INI file getting out of sync.

File: alembic/versions/85d52c4ca2ec_create_users_table.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        CREATE TABLE users (
+            id uuid DEFAULT uuid_generate_v4() PRIMARY KEY,
+            email VARCHAR,
+            dod_id VARCHAR,
+            first_name VARCHAR,
+            last_name VARCHAR
+        );
+    """)
+    db.execute("""
+        CREATE INDEX ix_users_email ON users (email);
```
Comment (jordangov) [inline alembic/versions/85d52c4ca2ec_create_users_table.py:32]: Should be unique, yes?

File: alembic/versions/85d52c4ca2ec_create_users_table.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    """)
+    db.execute("""
+        CREATE INDEX ix_users_dod_id ON users (dod_id);
```
Comment (jordangov) [inline alembic/versions/85d52c4ca2ec_create_users_table.py:35]: Should be unique, yes?

File: authnid/root.py [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+root = Blueprint('home', __name__)
+
+@root.route('/')
+def log_in_user():
+    # FIXME: Find or create user based on the X-Ssl-Client-S-Dn header
+    # TODO: Store/log the X-Ssl-Client-Cert in case it's needed?
+    if request.environ.get('HTTP_X_SSL_CLIENT_VERIFY') == 'SUCCESS' and is_valid_certificate(request):
+        sdn = request.environ.get('HTTP_X_SSL_CLIENT_S_DN')
+        # TODO: error handling for bad SDN, database errors, etc
```
Comment (jordangov) [inline authnid/root.py:13]: This is fine for now, but let's make sure we're tracking in PT as well.

**Review by jordangov — DISMISSED** [review]
Just the unique index thing, I think, then we should be good to go on this.

File: authnid/root.py (hunk 16 above)
Comment (dandds) [reply authnid/root.py:13]: sure: https://www.pivotaltracker.com/story/show/158725026

File: alembic/versions/85d52c4ca2ec_create_users_table.py (hunk 14 above)
Comment (dandds) [reply alembic/versions/85d52c4ca2ec_create_users_table.py:32]: 👍

File: alembic/versions/85d52c4ca2ec_create_users_table.py (hunk 15 above)
Comment (dandds) [reply alembic/versions/85d52c4ca2ec_create_users_table.py:35]: 👍

**Pull request merged**